### DISCUSSION TITLE
Building iPXE with HTTPS support

### DISCUSSION THREAD

**Discussion by Alex2357** [ipxe/ipxe#1357] [Q&A] [answered]
I created certs usign acme.sh (lets encrypt) for my home openwrt router.

In the end it reported 
```
[Fri Nov 29 18:xx:xx UTC 2024] The intermediate CA cert is in: /root/.acme.sh/somefqdn_ecc/ca.cer
[Fri Nov 29 18:xx:xx UTC 2024] And the full chain certs is there: /root/.acme.sh/somefqdn_ecc/fullchain.cer
```
I converted the certs to DER and configured uHTTPd. The Chrome reports certificate is valid if I open luci or attempt to download via https file https://somefqdn/isos/undionly.kpxe

Then I tried to do build of iPXE source https://github.com/ipxe/ipxe.git
Tried for commit 4f7dd7fbba205d413cf9b989f7cdc928fa02caf2 and for a3f1e8fb6707811e6eb90e339d7ebe813fd89a63

By running
```
make bin/ipxe.usb TRUST=$trustca CERT=$trustcert DEBUG=open,httpcore
```
where $trustca is the file of The intermediate CA /root/.acme.sh/somefqdn_ecc/ca.cer,
$trustcert is  the full chain certs is there: /root/.acme.sh/somefqdn_ecc/fullchain.cer
Note I use PEM certs(as documentation says) exactly as I get them from acme.sh, but not converted to DER certs.

The build completes ok.(For a3f1e8fb6707811e6eb90e339d7ebe813fd89a63 have to use distrobox image with fedora 32 and old  gcc, make etc, etc to avoid new compiler errors )

When I execute in KVM VM machine with content bin/ipxe.usb ( I run dd to write  into virtual hdd drive) I get next(I get same problem if I fash bin/ipxe.usb to USB and try on a real machine)

```
iPXE> imgfetch https://somefqdn/isos/undionly.kpxe

INTF Oxefe24+18 opening https URI
https://somefqdn/isos/undionly.kpxe...INTF Oxf0544+18 opening (SOCK_DGRAM,AF_UNKNOWN) socket
INTF ©xf01?4+50 opening (SOCK_STREAM,AF_INET) socket
Error: x420c6001 (http://ipxe.org/420c6001)
iPXE> imgfetch https://boot.ipxe.org/1mb
INTF @xef6b4+18 opening https URI
‘boot.ipxe.org/Imb...INTF Oxeffd4+18 opening (SOCK_DGRAM,AF_UNKNOWN) sock

INTF @xefca4+50 opening (SOCK_STREAM,AF_INET) socket
INTF @xef9?4+18 opening URI http://ca.ipxe.org/auto/c9Zeb62f.der?subject=MEBxCzA
JBGNUBAYTA LUTHSkuJuYDUQQKEYBIbnk lcm 14CBTZWN1cm] GeSBSZXN1YXU jaCBHem9 1cDEUMBMGA1U
EAXMMSUNSRYBSb2901F gx

INTF @xef974+18 opening http URI

INTF ©xf@bc4+*18 opening (SOCK_DGRAM,AF_UNKNOWN) socket

INTF ©xf@0a4+10 opening (SOCK_STREAM,AF_INET) socket

Permission denied (http://ipxe.org/0Z16eb3c)
```

What do I do wrong? How I can build it to make it work with HTTPS? I have enabled the proposed DEBUG options but nothing helps to understand what is going on.
It feels like download happens as it outputs dots like it downloads something before shows and error and then at the end some error appears.

**Comment by NiKiZe** [comment]
You do not have to embed the certificate in the client, what happens if you just build iPXE with Https enabled? And then, if you get an error read that URL and post the message and url?

**Reply by Alex2357** [reply]
@NiKiZe 

Short story - Build from sources without adding any certs - no luck. Wanted to provide proof that the cert is Trusted by Chrome but did not want to attach screenshot and decided to provide logs of wget. Find out both curl and wget do not trust cert. Reconfigured my home OpenWRT router to use nginx instead of uHTTPd. Received success. Both curl and wget trust the certificate and download correctly using TLSv1_2 and default TLSv1_3. iPXE Still doesn't work via https. Decided to try again and added certs by building using `make bin/ipxe.usb TRUST=$trustca CERT=$trustcert DEBUG=open,httpcore,tls` still doesn't work. Decided to exclude if somehow KVM affects and tried to use a regular machine. Flashed the bin/ipxe.usb to usb and tried on machine and got the same failure.

I was suspecting that iPXE is not using DNS provided by DHCP as result cert validation fails. But later I remembered when I had uHTTPd in which I disabled https redirect and allowed http. It worked usign http correctly with domain name. I just checked with NGINX and I get 302 redirect and then it attempts to go same host using https which tells me iPXE correctly resolves the ip. 
I still think it could be something wrong with DNS as the FQDN resolved by my openwrt to the openwrt server itself on the lan with local ip 192.168.x.x?
Could it be that somehow Lets Encrypt certs can not be validated? Is there a possibility to turn on more logging for TLS?

Long story steps in details of what I did.

## Creating distrobox VM with build tools and enter into it

```
distrobox create --name make-tools42 --image registry.fedoraproject.org/fedora-toolbox:42 --additional-packages "gcc perl binutils syslinux xz-devel openssl git"

distrobox enter make-tools42
```

Checking make installed

```
📦[someuser@make-tools42 src]$ make --version
GNU Make 4.4.1
Built for x86_64-redhat-linux-gnu
Copyright (C) 1988-2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
📦[someuser@make-tools42 src]$ 

```

Here's the source I have and what has modified to enable HTTPS

```
📦[someuser@make-tools42 src]$ git diff HEAD
diff --git a/src/config/general.h b/src/config/general.h
index 763a34aa..04823a42 100644
--- a/src/config/general.h
+++ b/src/config/general.h
@@ -56,7 +56,7 @@ FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
 
 #define        DOWNLOAD_PROTO_TFTP     /* Trivial File Transfer Protocol */
 #define        DOWNLOAD_PROTO_HTTP     /* Hypertext Transfer Protocol */
-#undef DOWNLOAD_PROTO_HTTPS    /* Secure Hypertext Transfer Protocol */
+#define        DOWNLOAD_PROTO_HTTPS    /* Secure Hypertext Transfer Protocol */
 #undef DOWNLOAD_PROTO_FTP      /* File Transfer Protocol */
 #undef DOWNLOAD_PROTO_SLAM     /* Scalable Local Area Multicast */
 #undef DOWNLOAD_PROTO_NFS      /* Network File System Protocol */
📦[someuser@make-tools42 src]$ git rev-parse HEAD
4f7dd7fbba205d413cf9b989f7cdc928fa02caf2
📦[someuser@make-tools42 src]$ 
```

## Building by running `make bin/ipxe.usb DEBUG=open,httpcore`

```
📦[someuser@make-tools42 src]$ make bin/ipxe.usb DEBUG=open,httpcore
  [DEPS] libgcc/__divdi3.c
  [DEPS] libgcc/__divmoddi4.c
  ...
  ...
  ...
  [BUILD] bin/ath9k.ids.o
  [AR] bin/blib.a
  [VERSION] bin/version.ipxe.hd.o
  [LD] bin/ipxe.hd.tmp
  [HOSTCC] util/zbin
  [BIN] bin/ipxe.hd.bin
  [ZINFO] bin/ipxe.hd.zinfo
  [ZBIN] bin/ipxe.hd.zbin
  [FINISH] bin/ipxe.hd
  [FINISH] bin/ipxe.usb
rm bin/ipxe.hd.zinfo bin/ipxe.hd.bin bin/ipxe.hd.zbin bin/version.ipxe.hd.o bin/ipxe.hd
📦[someuser@make-tools42 src]$ 
```

## Creating hdd for KVM 

```
usbimgpath=/tmp/usbimghdd.img
qemu-img create -f raw $usbimgpath 10M
dd if=bin/ipxe.usb of=$usbimgpath bs=4M conv=notrunc
```

Log of creating HDD

```
usbimgpath=/tmp/usbimghdd.img
qemu-img create -f raw $usbimgpath 10M
dd if=bin/ipxe.usb of=$usbimgpath bs=4M conv=notrunc
Formatting '/tmp/usbimghdd.img', fmt=raw size=10485760
0+1 records in
0+1 records out
409600 bytes (410 kB, 400 KiB) copied, 0.000608519 s, 673 MB/s
```

## Creating and starting KVM VM to test

sudo virt-install \
  --name ipxe-test \
  --memory 2560 \
  --vcpus 1 \
  --os-variant generic \
  --network network=default,model=virtio \
  --disk path=$usbimgpath,device=disk,bus=sata,format=raw \
  --boot hd \
  --import \
  --noautoconsole

## Result - Fail

```
iPXE (https://ipxe.org) 00:03.0 CA00 PCI2.10 PnP PMM+9EFCBE60+9EFOBE60 CA00

Booting from Hard Disk...
iPXE initialising devices...

iPXE 1.21.1+ (g4f7d) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP HTTPS iSCSI TFTP SRP AoE ELF MBOOT PXE bzImage Menu PXEXT

iPXE> dhcp
INTF 0xe4d894+18 opening (SOCK_DGRAM,AF_INET) socket
Configuring (net0 52:54:00:7e:eb:d2)...... ok

iPXE> imgfetch https://z3.subdtest.somedomain.com.au/undionly.kpxe
INTF 0xf1924+18 opening HTTPS URI
https://z3.subdtest.somedomain.com.au/undionly.kpxe...INTF 0xf5b44+18 opening (SOCK_DGRAM,AF_UNKNOWN) socket
INTF 0xf5b44+50 opening (SOCK_STREAM,AF_INET) socket
.......................... Error 0x420c6001 (https://ipxe.org/420c6001)
iPXE>

```

## Checking using wget - The cert is not trusted, while Chrome trust. Suspect `uHTTPd` incorrectly returns the full chain or I made mistake of converting to DER as `uHTTPd` requires DER. 

```
📦[someuser@make-tools42 src]$ wget --secure-protocol=TLSv1_2 https://z3.subdtest.somedomain.com.au/undionly.kpxe
ERROR: The certificate is NOT trusted. The certificate issuer is unknown. 

                          [Files: 0  Bytes: 0  [0 B/s] Redirects: 0  Todo: 0  Errors: 0                                    ]
📦[someuser@make-tools42 src]$ 
📦[someuser@make-tools42 src]$ wget --no-check-certificate https://z3.subdtest.somedomain.com.au/undionly.kpxe
WARNING: The certificate is NOT trusted. The certificate issuer is unknown. 
undionly.kpxe        100% [===============================================================================================>]   68.52K    --.-KB/s
                          [Files: 1  Bytes: 68.52K [1.31MB/s] Redirects: 0  Todo: 0  Errors: 0                             ]
📦[someuser@make-tools42 src]$ 

```

wget --no-proxy --secure-protocol=TLSv1_2 --user-agent="Mozilla/5.0" https://z3.subdtest.somedomain.com.au/undionly.kpxe

```
📦[someuser@make-tools42 src]$ wget --no-proxy --secure-protocol=TLSv1_2 --user-agent="Mozilla/5.0" https://z3.subdtest.somedomain.com.au/undionly.kpxe
ERROR: The certificate is NOT trusted. The certificate issuer is unknown. 

                          [Files: 0  Bytes: 0  [0 B/s] Redirects: 0  Todo: 0  Errors: 0                                    ]
📦[someuser@make-tools42 src]$ 
```




## WGET & CURL both work with NGINX on OpenWRT correctly

The nginx server serves correctly so that both wget & curl trust the cert and both show it can work TLSv1_2 and TLSv1_3. 

```
📦[someuser@make-tools42 src]$ wget  --secure-protocol=TLSv1_2  https://z3.subdtest.somedomain.com.au/undionly.kpxe
undionly.kpxe.1      100% [===============================================================================================>]   68.52K    --.-KB/s
                          [Files: 1  Bytes: 68.52K [951.79KB/s] Redirects: 0  Todo: 0  Errors: 0                           ]
📦[someuser@make-tools42 src]$ 

📦[someuser@make-tools42 src]$ curl --tlsv1.2 --tls-max 1.2   https://z3.subdtest.somedomain.com.au/undionly.kpxe --output undionly.kpxe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 70174  100 70174    0     0  2939k      0 --:--:-- --:--:-- --:--:-- 2979k
📦[someuser@make-tools42 src]$ curl    https://z3.subdtest.somedomain.com.au/undionly.kpxe --output undionly.kpxe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 70174  100 70174    0     0  3240k      0 --:--:-- --:--:-- --:--:-- 3263k
📦[someuser@make-tools42 src]$ curl -v   https://z3.subdtest.somedomain.com.au/undionly.kpxe --output undionly.kpxe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host z3.subdtest.somedomain.com.au:443 was resolved.
* IPv6: (none)
* IPv4: 192.168.4.1
*   Trying 192.168.4.1:443...
* ALPN: curl offers h2,http/1.1
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
*  CAfile: /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
*  CApath: none
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [25 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [2955 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [80 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384 / x25519 / id-ecPublicKey
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: CN=z3.subdtest.somedomain.com.au
*  start date: Nov 29 00:00:00 2024 GMT
*  expire date: Feb 27 23:59:59 2025 GMT
*  subjectAltName: host "z3.subdtest.somedomain.com.au" matched cert's "z3.subdtest.somedomain.com.au"
*  issuer: C=AT; O=ZeroSSL; CN=ZeroSSL ECC Domain Secure Site CA
*  SSL certificate verify ok.
*   Certificate level 0: Public key type EC/prime256v1 (256/128 Bits/secBits), signed using ecdsa-with-SHA384
*   Certificate level 1: Public key type EC/secp384r1 (384/192 Bits/secBits), signed using ecdsa-with-SHA384
*   Certificate level 2: Public key type EC/secp384r1 (384/192 Bits/secBits), signed using ecdsa-with-SHA384
* Connected to z3.subdtest.somedomain.com.au (192.168.4.1) port 443
* using HTTP/1.x
} [5 bytes data]
> GET /undionly.kpxe HTTP/1.1
> Host: z3.subdtest.somedomain.com.au
> User-Agent: curl/8.10.1
> Accept: */*
> 
* Request completely sent off
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [281 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [281 bytes data]
< HTTP/1.1 200 OK
< Server: nginx/1.25.0
< Date: Sun, 08 Dec 2024 01:23:46 GMT
< Content-Type: application/octet-stream
< Content-Length: 70174
< Last-Modified: Thu, 28 Nov 2024 09:39:42 GMT
< Connection: keep-alive
< ETag: "67483a5e-1121e"
< Accept-Ranges: bytes
< 
{ [16127 bytes data]
100 70174  100 70174    0     0  3181k      0 --:--:-- --:--:-- --:--:-- 3263k
* Connection #0 to host z3.subdtest.somedomain.com.au left intact
📦[someuser@make-tools42 src]$ 


```

## Still doesn't work even with nginx but I find out interesting thing DEBUG=tls in https://ipxe.org/410de13c

```
iPXE 1.21.1+ (g4f7ad) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP HTTPS iSCSI TFTP SRP AoE ELF MBOOT PXE bzImage Menu PXEXT

net0: 52:54:00:7e:eb:d2 using virtio-net on 0000:00:03.0 (Ethernet) [open]
  [Link:up, TX:0 TXE:0 RX:0 RXE:0]

INT 0xe4854+18 opening (SOCK_DGRAM,AF_INET) socket
Configuring (net0 52:54:00:7e:eb:d2)...... ok
net0: 192.168.122.57/255.255.255.0 gw 192.168.122.1
Nothing to boot: No such file or directory (https://ipxe.org/2d03e13b)
No more network devices

iPXE> ifstat
net0: 52:54:00:7e:eb:d2 using virtio-net on 0000:00:03.0 (Ethernet) [open]
  [Link:up, TX:9 TXE:0 RX:34 RXE:311]
RXE: 7 x "Operation not supported (https://ipxe.org/3c3f6303)"
RXE: 18 x "Operation not supported (https://ipxe.org/3c086003)"
RXE: 6 x "The socket is not connected (https://ipxe.org/380f6001)"

iPXE> imgfetch https://z3.subdtest.somedomain.com.au/undionly.kpxe
INT 0xf18e4+18 opening HTTPS URI
https://z3.subdtest.somedomain.com.au/undionly.kpxe... INT 0xf5b04+18 opening (SOCK_DGRAM,AF_UNKNOWN) socket
INT 0x5874+50 opening (SOCK_STREAM,AF_INET) socket
Operation not permitted (https://ipxe.org/410de13c)

```

## Repeating with DEBUG=open,httpcore,tls

make bin/ipxe.usb DEBUG=open,httpcore,tls

📦[someuser@make-tools42 src]$ make clean
rm -f -r bin/*.* bin-*/*.* bin/.certificate.* bin-*/.certificate.* bin/.certificates.* bin-*/.certificates.* bin/.private_key.* bin-*/.private_key.* bin/errors bin-*/errors bin/NIC bin-*/NIC ./util/zbin ./util/zbin ./util/elf2efi32 ./util/elf2efi32 ./util/elf2efi64 ./util/elf2efi64 ./util/efirom ./util/efirom ./util/efifatbin ./util/efifatbin ./util/einfo ./util/einfo TAGS TAGS bin/symtab bin-*/symtab
📦[someuser@make-tools42 src]$

Runing `make bin/ipxe.usb DEBUG=open,httpcore,tls`



[someuser@fedora-sb src]$ dd if=bin/ipxe.usb of=$usbimgpath bs=4M conv=notrunc
0+1 records in
0+1 records out
409600 bytes (410 kB, 400 KiB) copied, 0.0106225 s, 38.6 MB/s
[someuser@fedora-sb src]$ 

## Still fails. Not much info. I think I need to specify cert as I did earlier.



```
net0: 192.168.122.57/255.255.255.0 gw 192.168.122.1
Nothing to boot: No such file or directory (https://ipxe.org/2d03e13b)
No more network devices

iPXE> dhcp
INT 0xf6a94+18 opening (SOCK_DGRAM,AF_INET) socket
Configuring (net0 52:54:00:7e:eb:d2)...... ok

iPXE> ifstat
ifstat: command not found

iPXE> ifstat
net0: 52:54:00:7e:eb:d2 using virtio-net on 0000:00:03.0 (Ethernet) [open]
  [Link:up, TX:13 TXE:0 RX:51 RXE:451]
RXE: 22 x "Operation not supported (https://ipxe.org/3c086003)"
RXE: 17 x "Operation not supported (https://ipxe.org/3c3f6303)"
RXE: 6 x "The socket is not connected (https://ipxe.org/380f6001)"

iPXE> imgfetch https://z3.subdtest.somedomain.com.au/undionly.kpxe
INT 0xf6414+18 opening HTTPS URI
TLS 0xfbb24 created session z3.subdtest.somedomain.com.au
https://z3.subdtest.somedomain.com.au/undionly.kpxe... INT 0xfa74+18 opening (SOCK_DGRAM,AF_UNKNOWN) socket
INT 0xfbb24+50 opening (SOCK_STREAM,AF_INET) socket
TLS 0xfbb24 received fatal alert 40
Operation not permitted (https://ipxe.org/410de13c)

```

## Specified Certs - Still no luck

trustca=/var/home/someuser/ca.crt 
trustcert=/var/home/someuser/fullchain-pem-z3.subdtest.somedomain.com.au.crt

echo "trustca=$trustca    trustcert=$trustcert"
sha256sum $trustca $trustcert

```
📦[someuser@make-tools42 src]$ trustca=/var/home/someuser/ca.crt 
trustcert=/var/home/someuser/fullchain-pem-z3.subdtest.somedomain.com.au.crt

echo "trustca=$trustca    trustcert=$trustcert"
sha256sum $trustca $trustcert
trustca=/var/home/someuser/ca.crt    trustcert=/var/home/someuser/fullchain-pem-z3.subdtest.somedomain.com.au.crt
32...04  /var/home/someuser/ca.crt
02...99  /var/home/someuser/fullchain-pem-z3.subdtest.somedomain.com.au.crt
📦[someuser@make-tools42 src]$ 

```


make bin/ipxe.usb TRUST=$trustca CERT=$trustcert DEBUG=open,httpcore,tls

## Validating that this is not KVM issue - Getting exactly same results from a regular machine as in KVM


sudo dd if=bin/ipxe.usb of=/dev/sda

**Reply by NiKiZe** [reply]
If the connection is working in Firefox you do generally not need to embed anything unless you have an locked down network that doesn't allow connections to fetch certificates from iPXE servers, or blocking some other checks.

So again let's make this simple if you run iPXE without modifications (only enabling Https) you get permission denied with the https://ipxe.org/410de13c url?

And with DEBUG=tls you do get `TLS 0xfbb24 received fatal alert 40` ? (Your output above with this debug does probably not actually have that debug enabled)

Do consider using a packet capture to see if that yields better insight in to what might go wrong with the tls handshake, not ruling out incompatible ciphers.

**Reply by Alex2357** [reply]
After googling I find out that TLS 0xfbb24 received fatal alert 40 is TLSv1.1. After doing check I got next errors which confirm my openwrt doesn't support it

openssl s_client -connect z3.....com.au:443 -tls1_1
Connecting to 192.168.x.1
CONNECTED(00000003)
C0528D6AAD7F0000:error:0A0000BF:SSL routines:tls_setup_handshake:no protocols available:ssl/statem/statem_lib.c:153:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 7 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---

**Reply by Alex2357** [reply]
I believe that iPXE connects using TLSv1.1 it might be something needs to be defined to support TLSv1.2 or this https://ipxe.org/crypto page is incorrect in saying it supports TLSv1.2

**Comment by mcb30** [comment, marked as answer]
@Alex2357 You are attempting to connect to a site that has an ECC server certificate.  ECC support is not yet present in iPXE (but is being actively worked on).

Assuming that the site you are trying to connect to is your own site, then you can fix the problem by using an RSA server certificate instead.

As @NiKiZe says, you do not need to use any custom `TRUST=...` or `CERT=...` lines when building iPXE.

**Reply by Alex2357** [reply]
Thank you very much for letting me know about ECC. Will try to request RSA and see how I go.

**Reply by Alex2357** [reply]
Thanks to everyone for your help. In the end it worked when I used RSA keys.
I now face another issue with creating a menu. But I hope I will figure out by myself.

**Reply by Alex2357** [reply]
Closing as it works for RSA.
